### PR TITLE
Git credentials

### PR DESCRIPTION
**Pull request by LysandreJik**
*Thoroughly tested in practice, will implement unit tests in a bit but the API will remain as it is so pinging you for review.*

Fixes https://github.com/huggingface/huggingface_hub/issues/211 and https://github.com/huggingface/moon-landing/issues/150

#### Problem

This PR offers a long-awaited overhaul of the login system through the `huggingface_hub` library. The way the login system is currently designed is complex and led to a lot of errors mid-training during the Jax/Flax community week - as well as a general misunderstanding of what was happening behind the scenes.

The core issue is that logging in through `huggingface_cli login` does *not* log the user through git. The user will have to reauthenticate when pushing. This is particularly an issue on Google Colab, where a simple `git push` is met by a fatal error:
```
fatal: could not read Username for 'https://huggingface.co': No such device or address
```
The workarounds are cumbersome (see [#12572](https://github.com/huggingface/transformers/issues/12572)).

#### Solving the issue

This PR offers to use the `git-credential` API to store the username and password in a way that git can leverage when authenticating the user on `huggingface.co`. It leverages the `git credential.helper store` helper, which stores the username/password under plain text in a file with the user's permission.

It does not set the global credential helper to `store` as this could disrupt the user's git configuration - it instead detects what is the current helper, and if it isn't `store`, mentions that users will need to reauthenticate. If users do not wish to reauthenticate, a command to be run (which will set the helper to `store`) is printed out.

Additionally, in order to improve the experience across platforms, this PR adjusts two configuration attributes which are linked to the above:
- Cloning a repository through `Repository(xxx, clone_from=yyy)` will set the helper of that repository to `store`. 
- Using the `Repository` without explicitly passing a `git_email` or a `git_user`, instead of leveraging the system's default, will leverage the token to be used by that `Repository`.

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Very clean! Thanks for investigating that bug and finding this solution!

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -224,6 +225,80 @@ def __str__(self):
...
+
+        standard_input += "\n"
+        print(standard_input.encode("utf-8"))
+
+        process.stdin.write(standard_input.encode("utf-8"))
+        process.stdin.flush()
+        output = process.stdout.read()
+        print(output)
+        output = output.decode("utf-8")
+        print(output)
```
Comment (osanseviero) [inline src/huggingface_hub/hf_api.py:270]: Should we remove these `print` statements?

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -240,6 +315,9 @@ def login(self, username: str, password: str) -> str:
         r = requests.post(path, json={"username": username, "password": password})
         r.raise_for_status()
         d = r.json()
+
+        cased_username = self.whoami(d["token"])["name"]
```
Comment (osanseviero) [inline src/huggingface_hub/hf_api.py:319]: No strong opinion.

is this the same as `username.lower()`? Maybe we could save an API call here, but maybe this is safer to be consistent with `erase_from_credential_store`

**Review by osanseviero — APPROVED** [review]
Thanks for this PR, this is a great improvement of quality of life for development!

File: src/huggingface_hub/commands/user.py [hunk 3]
```diff
@@ -117,6 +117,27 @@ def tabulate(rows: List[List[Union[str, int]]], headers: List[str]) -> str:
...
+            stdout=subprocess.PIPE,
+            encoding="utf-8",
+            check=True,
+            cwd=directory,
+        ).stdout.split("\n")
+
+        current_credential_helper = ""
+        for line in output:
+            if "credential.helper" in line:
+                current_credential_helper = line.split("=")[-1]
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply src/huggingface_hub/commands/user.py]: Indeed, I instead used an array of helpers. Thank you!

File: src/huggingface_hub/hf_api.py (hunk 2 above)
Comment (LysandreJik) [reply src/huggingface_hub/hf_api.py]: Absolutely! I went ahead the "correctly" cased way - but any casing works as long as it's identical across the helper methods. Thanks!

**Pull request merged**